core/services/relay/evm/chain_reader.go

```diff
@@ -28,31 +34,46 @@ func newChainReader(lggr logger.Logger, chain evm.Chain, ropts *types.RelayOpts)
 		return nil, relaytypes.ErrorChainReaderUnsupported{}
 	}
 
-	if err = ValidateChainReaderConfig(*relayConfig.ChainReader); err != nil {
+	if err = parseChainContractReadersABIs(relayConfig.ChainReader.ChainContractReaders); err != nil {
+		return nil, err
+	}
+
+	if err = validateChainReaderConfig(*relayConfig.ChainReader); err != nil {
 		return nil, fmt.Errorf("invalid ChainReader configuration: %w", err)
 	}
 
-	return NewChainReaderService(lggr, chain.LogPoller())
+	return &chainReader{lggr.Named("ChainReader"), chain.LogPoller(), relayConfig.ChainReader.ChainContractReaders, chain}, nil
 }
 
-func ValidateChainReaderConfig(cfg types.ChainReaderConfig) error {
-	for contractName, chainContractReader := range cfg.ChainContractReaders {
-		abi, err := abi.JSON(strings.NewReader(chainContractReader.ContractABI))
+func parseChainContractReadersABIs(chainContractReaders map[string]types.ChainContractReader) error {
+	for key, chainContractReader := range chainContractReaders {
+		parsedABI, err := abi.JSON(strings.NewReader(chainContractReader.ContractABI))
 		if err != nil {
-			return err
+			return fmt.Errorf("falied to parse contract:%s abi:%s, err:%w", key, chainContractReader.ContractABI, err)
+		}
+		chainContractReader.ParsedContractABI = &parsedABI
+		chainContractReaders[key] = chainContractReader
+	}
+	return nil
+}
+
+func validateChainReaderConfig(cfg types.ChainReaderConfig) (err error) {
+	for contractName, chainContractReader := range cfg.ChainContractReaders {
+		if chainContractReader.ParsedContractABI == nil {
+			return fmt.Errorf("contract: %s ABI is not parsed", contractName)
 		}
 
 		for chainReadingDefinitionName, chainReaderDefinition := range chainContractReader.ChainReaderDefinitions {
 			switch chainReaderDefinition.ReadType {
 			case types.Method:
-				err = validateMethods(abi, chainReaderDefinition)
+				err = validateMethods(*chainContractReader.ParsedContractABI, chainReaderDefinition)
 			case types.Event:
-				err = validateEvents(abi, chainReaderDefinition)
+				err = validateEvents(*chainContractReader.ParsedContractABI, chainReaderDefinition)
 			default:
 				return fmt.Errorf("invalid chain reader definition read type: %d", chainReaderDefinition.ReadType)
 			}
 			if err != nil {
-				return fmt.Errorf("invalid chain reader config for contract: %q chain reading definition: %q, err: %w", contractName, chainReadingDefinitionName, err)
+				return errors.Wrap(err, fmt.Sprintf("invalid chain reader config for contract: %q chain reading definition: %q", contractName, chainReadingDefinitionName))
```

Comment:
As I understand it, the recommended replacement for `errors.Wrap(err, "...")` is `fmt.Errorf("...: %w" err)`.  Was there a problem caused by using `fmt` here?

Reply:
changed to fmt